src/WinRMSharp/Utils/Xml.cs

```diff
@@ -6,15 +6,72 @@ namespace WinRMSharp.Utils
 {
     internal static class Xml
     {
+        private static readonly Dictionary<string, string> NamespaceToPrefix = new()
+        {
+            { Namespace.XML_SCHEMA, "xsd" },
+            { Namespace.XML_SCHEMA_INSTANCE, "xsi" },
+            { Namespace.SOAP_ENVELOPE, "env" },
+            { Namespace.ADDRESSING, "a" },
+            { Namespace.CIM_BINDING, "b" },
+            { Namespace.ENUMERATION, "n" },
+            { Namespace.TRANSFER, "x" },
+            { Namespace.WSMID, "wsmid" },
+            { Namespace.WSMAN_0, "w" },
+            { Namespace.WSMAN_1, "p" },
+            { Namespace.WSMAN_SHELL, "rsp" },
+            { Namespace.WSMAN_CONFIG, "cfg"}
+        };
+
         public static string Serialize<T>(T obj)
         {
-            XmlSerializer serializer = new XmlSerializer(typeof(Envelope));
+            // First pass: Generate temporary XML to detect used namespaces
+            string tempXml = SerializeTemp(obj);
+
+            // Extract namespaces actually used in the XML
+            HashSet<string> usedNamespaces = ExtractNamespaces(tempXml);
 
-            using (StringWriter textWriter = new StringWriter())
+            // Build XmlSerializerNamespaces with desired prefixes
+            XmlSerializerNamespaces namespaces = new XmlSerializerNamespaces();
+            foreach (string nsUri in usedNamespaces)
             {
-                serializer.Serialize(textWriter, obj);
-                return textWriter.ToString();
+                string? prefix = NamespaceToPrefix[nsUri];
+                namespaces.Add(prefix, nsUri);
             }
+
+            // Second pass: Serialize with filtered namespaces
+            return SerializeFinal(obj, namespaces);
+        }
+
+        private static string SerializeTemp<T>(T obj)
+        {
+            XmlSerializer serializer = new XmlSerializer(typeof(T));
+            using StringWriter writer = new StringWriter();
+            serializer.Serialize(writer, obj);
+            return writer.ToString();
+        }
+
+        private static string SerializeFinal<T>(T obj, XmlSerializerNamespaces namespaces)
+        {
+            XmlSerializer serializer = new XmlSerializer(typeof(T));
+            using StringWriter writer = new StringWriter();
+            serializer.Serialize(writer, obj, namespaces);
+            return writer.ToString();
+        }
+
+        private static HashSet<string> ExtractNamespaces(string xml)
+        {
+            XDocument doc = XDocument.Parse(xml);
+            HashSet<string> namespaces = new HashSet<string>();
+
+            foreach (XElement? element in doc.Descendants())

```

Comment:
Is `element` & `attr` actually nullable here? Won't that cause issues when you try and access `element.Attributes()` and `attr.Value`?